core/drivers/crypto/caam/caam_ctrl.c

```diff
@@ -81,6 +73,20 @@ static TEE_Result crypto_driver_init(void)
 
 early_driver_init(crypto_driver_init);
```

Comment:
Is it something really necessary? crypto_driver_init can directly called in the tee_cryp_init function (see below) that enabled the crypto before any crypto operations (like HUK generation). The concept is to be able to replace the CAAM driver with another HW crypto driver easly. 

__weak TEE_Result crypto_driver_init(void)
{
	return TEE_SUCCESS;
}


static TEE_Result tee_cryp_init(void)
{
	TEE_Result res = crypto_init();

	/*
	 * If there is a Cryptographic Driver, we need to
	 * call the initialization function here before using
	 * cryptographic operation during the boot
	 * (e.g. HUK generation)
	 */
	if (res == TEE_SUCCESS)
		res = crypto_driver_init();

	if (res) {
		EMSG("Failed to initialize crypto API: %#" PRIx32, res);
		panic();
	}

	plat_rng_init();

	return TEE_SUCCESS;
}

Reply:
I prefer the clarity of using an explicit new initcall which runs before services are initialized. This also allows a clean handover of jobrings when `CFG_NXP_CAAM_RUNTIME_JR=n`.

Reply:
I would let core maintainers to comment and adjudicated on this point.
Using initcall never guaranty the execution order in case of concurrency.

Reply:
@cneveux what do you mean "concurrency"? initcalls do provide some execution order guarantees between different levels, which should be enough in this case?

As to initcalls vs. direct calls, no strong opinion but I tend to prefer the first because they are bit simpler: no weak symbol, no need to include headers and insert explicit calls, each layer remains independent on the others.

Reply:
@jforissier 
My concern is about the initcalls order which raised each time we need to init A before B and finally C because without A, B can not be initialized and C can not work without B. 
In this case, C can be init thru driver_init and to ensure that B is initialized before C, B is init thru early_driver_init. How to ensure that A is initialized first, we will add a early_early_driver_init?

Reply:
It's a matter of having the "right" number of levels. Dismissing the mechanism just because it doesn't scale perfectly well to large numbers is a bit excessive I think. All the more since it's just a matter of naming (initcall levels are just integers in the end).

Currently we have "services" which are initialized first and then "drivers". The current use case shows that a service (crypto) may depend on specific hardware and need a driver. So introducing "early drivers" makes sense. We don't have any use case where such an "early driver" would depend on another driver yet. So your argument is valid in theory, but is it in practice?

Reply:
I will not debate on this point. I'm not rejecting the current proposal, for sure today solution is acceptable. I'm just saying that we may face another problem something later.